test/task_graph_tests.cc

```diff
@@ -0,0 +1,448 @@
+		auto tid_collective_explicit_2 = test_utils::add_host_task(tm, experimental::collective(group), [](handler&) {});
+
+		CHECK_FALSE(has_any_dependency(tm, tid_master, tid_collective_implicit_1));
+		CHECK_FALSE(has_any_dependency(tm, tid_master, tid_collective_implicit_2));
+		CHECK_FALSE(has_any_dependency(tm, tid_master, tid_collective_explicit_1));
+		CHECK_FALSE(has_any_dependency(tm, tid_master, tid_collective_explicit_2));
+
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_implicit_1, tid_master));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_implicit_1, tid_collective_implicit_2));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_implicit_1, tid_collective_explicit_1));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_implicit_1, tid_collective_explicit_2));
+
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_implicit_2, tid_master));
+		CHECK(has_dependency(tm, tid_collective_implicit_2, tid_collective_implicit_1, dependency_kind::ORDER_DEP));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_implicit_2, tid_collective_explicit_1));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_implicit_2, tid_collective_explicit_2));
+
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_explicit_1, tid_master));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_explicit_1, tid_collective_implicit_1));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_explicit_1, tid_collective_implicit_2));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_explicit_1, tid_collective_explicit_2));
+
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_explicit_2, tid_master));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_explicit_2, tid_collective_implicit_1));
+		CHECK_FALSE(has_any_dependency(tm, tid_collective_explicit_2, tid_collective_implicit_2));
+		CHECK(has_dependency(tm, tid_collective_explicit_2, tid_collective_explicit_1, dependency_kind::ORDER_DEP));
+	}
+
+	void check_path_length_and_front(task_manager& tm, int path_length, std::unordered_set<task_id> exec_front) {
+		{
+			INFO("path length");
+			CHECK(task_manager_testspy::get_max_pseudo_critical_path_length(tm) == path_length);
+		}
+		{
+			INFO("execution front");
+			std::unordered_set<task*> task_exec_front;
+			std::transform(exec_front.cbegin(), exec_front.cend(), std::inserter(task_exec_front, task_exec_front.begin()),
+			    [&tm](task_id tid) { return const_cast<task*>(tm.get_task(tid)); });
+			CHECK(task_manager_testspy::get_execution_front(tm) == task_exec_front);
+		}
+	}
+
+	TEST_CASE("task_manager keeps track of max pseudo critical path length and task front", "[task_manager][task-graph][task-front]") {
+		using namespace cl::sycl::access;
+		task_manager tm{1, nullptr, nullptr};
+		test_utils::mock_buffer_factory mbf(&tm);
+		auto buf_a = mbf.create_buffer(cl::sycl::range<1>(128));
+
+		SECTION("with true dependencies") {
```

Comment:
Should this also test anti- and order-dependencies?

Reply:
I thought about it, but none of the horizon code ever looks at the *types* of dependencies, so it's probably not necessary. I'd keep it open for future expansion.

Reply:
I see. In that case I'd advocate removing the section altogether, as it's existence seems to imply there *is* some significance to the type of dependencies.

Reply:
Agreed,